### ISSUE TITLE
Problem using `execute` and boolean parameters

### ISSUE THREAD

**Issue by joakimbeng** [sidorares/node-mysql2#446]
First of all, great job with this module! It was a really smooth transition from the `mysql` module.

Now to the problem...

I have a table with a `isAddon TINYINT(1) NOT NULL` column.

Such a column is converted to a `boolean` when selecting from the table, but when using a `boolean` parameter to filter the table only works with the `query()` method and not `execute()`. I get the same result no matter if I pass `true` or `false` as a parameter, see example below.

**This does not work:**

```javascript
// running a query where I want: isAddon=0
db.execute('select count(*) from MyTable where isAddon = ?', [false]).then(console.log)
Promise { <pending> }
> [ [ BinaryRow { 'count(*)': 680 } ],
  [ { catalog: 'def',
      schema: '',
      name: 'count(*)',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 63,
      columnLength: 21,
      columnType: 8,
      flags: 129,
      decimals: 0 } ] ]

// running a query where I want: isAddon=1
> db.execute('select count(*) from MyTable where isAddon = ?', [true]).then(console.log)
Promise { <pending> }
> [ [ BinaryRow { 'count(*)': 680 } ],
  [ { catalog: 'def',
      schema: '',
      name: 'count(*)',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 63,
      columnLength: 21,
      columnType: 8,
      flags: 129,
      decimals: 0 } ] ]
```

As you can see the results are the same, I get `680` number of rows for both queries.

**Using `query()` works:**

```javascript
// running the query where I want: isAddon=0
> db.query('select count(*) from MyTable where isAddon = ?', [false]).then(console.log)
Promise { <pending> }
> [ [ TextRow { 'count(*)': 680 } ],
  [ { catalog: 'def',
      schema: '',
      name: 'count(*)',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 63,
      columnLength: 21,
      columnType: 8,
      flags: 129,
      decimals: 0 } ] ]

// running the query where I want: isAddon=1
> db.query('select count(*) from MyTable where isAddon = ?', [true]).then(console.log)
Promise { <pending> }
> [ [ TextRow { 'count(*)': 7 } ],
  [ { catalog: 'def',
      schema: '',
      name: 'count(*)',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 63,
      columnLength: 21,
      columnType: 8,
      flags: 129,
      decimals: 0 } ] ]
```

As you can see I get different results, which are correct, `680` number of rows where `isAddon=0` and `7` rows where `isAddon=1`.

**Converting `boolean`'s to `number`'s also work with `execute()`:**

```javascript
> db.execute('select count(*) from MyTable where isAddon = ?', [Number(false)]).then(console.log)
Promise { <pending> }
> [ [ BinaryRow { 'count(*)': 680 } ],
  [ { catalog: 'def',
      schema: '',
      name: 'count(*)',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 63,
      columnLength: 21,
      columnType: 8,
      flags: 129,
      decimals: 0 } ] ]

> db.execute('select count(*) from MyTable where isAddon = ?', [Number(true)]).then(console.log)
Promise { <pending> }
> [ [ BinaryRow { 'count(*)': 7 } ],
  [ { catalog: 'def',
      schema: '',
      name: 'count(*)',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 63,
      columnLength: 21,
      columnType: 8,
      flags: 129,
      decimals: 0 } ] ]
```

---

NodeJS version: `v6.7.0`  
mysql2 version: `1.1.1`

**Comment by sidorares** [comment]
hi @joakimbeng  - would you be able to post here very simple self contained example ( creating temporary table, inserting few records and doing select )?

Also if you can post output with `debug` connection option that would be useful as well

`query()` and `execute()` parameters work very differently internally, despite similar user facing api. query sends sql as is after formatting values ( you can see result by looking at `Query` object - returned by non-promise version: `var q = conn.query('select 1 + ?', [2]); console.log(q.sql)` ). Execute does `prepare` first for non-formatted sql `select 1 + ?` (if not previously prepared)  and then sends parameter in a separate packet.

**Comment by joakimbeng** [comment]
Ok, here's a contained example:

**The data:**

```javascript
conn.query(`
  create temporary table my_table (
    val varchar(32) not null,
    bit_prop tinyint(1) not null
  );
  insert into my_table (val, bit_prop)
  values ('lorem', 0),
         ('ipsum', 1),
         ('dolor', 1),
         ('sit', 0),
         ('amet', 1);
`)
```

**The query:**

```javascript
conn.execute(`select * from my_table where bit_prop = ?`, [true], console.log)
```

**The full output:**

<details>
```javascript
Add command: execute
Add command: execute
2 4777 <== execute#start(0,,23)
2 4777 <== 13000000170100000000010000000001fd000474727565
Execute {
  domain:
   Domain {
     domain: null,
     _events: { error: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] },
  _events: {},
  _eventsCount: 0,
  _maxListeners: undefined,
  next: [Function],
  statement:
   PreparedStatementInfo {
     query: 'select * from my_table where bit_prop = ?',
     id: 1,
     columns: [ [Object], [Object] ],
     parameters: [ [Object] ],
     rowParser: null,
     _connection:
      Connection {
        domain: [Object],
        _events: {},
        _eventsCount: 0,
        _maxListeners: undefined,
        config: [Object],
        stream: [Object],
        _internalId: 2,
        _commands: [Object],
        _command: [Circular],
        _paused: false,
        _paused_packets: [Object],
        _statements:
         LRUCache {
           max: 16000,

           'undefined/undefined/undefinedselect * from my_table where bit_prop = ?' => { value: [Circular] }
         },
        textProtocolParsers: {},
        binaryProtocolParsers: [Object],
        serverCapabilityFlags: 2148530175,
        authorized: true,
        sequenceId: 1,
        threadId: 4777,
        _handshakePacket: [Object],
        _fatalError: null,
        _protocolError: null,
        _outOfOrderPackets: [],
        packetParser: [Object],
        serverEncoding: 'cesu8',
        connectTimeout: null,
        connectionId: 4777 } },
  sql: 'select * from my_table where bit_prop = ?',
  values: [ true ],
  onResult:
   { [Function: runBound]
     domain:
      Domain {
        domain: null,
        _events: [Object],
        _eventsCount: 1,
        _maxListeners: undefined,
        members: [] } },
  parameters: [ true ],
  insertId: 0,
  _rows: [],
  _fields: [],
  _result: [],
  _fieldCount: 0,
  _rowParser: null,
  _executeOptions:
   { sql: 'select * from my_table where bit_prop = ?',
     values: [ true ] },
  _resultIndex: 0,
  _localStream: null,
  _unpipeStream: [Function],
  _streamFactory: undefined,
  _connection:
   Connection {
     domain:
      Domain {
        domain: null,
        _events: [Object],
        _eventsCount: 1,
        _maxListeners: undefined,
        members: [] },
     _events: {},
     _eventsCount: 0,
     _maxListeners: undefined,
     config:
      ConnectionConfig {
        isServer: undefined,
        stream: undefined,
        host: '172.17.0.2',
        port: '3306',
        localAddress: undefined,
        socketPath: undefined,
        user: 'admin',
        password: 'admin',
        passwordSha1: undefined,
        database: 'my_database',
        connectTimeout: 10000,
        insecureAuth: false,
        supportBigNumbers: false,
        bigNumberStrings: false,
        decimalNumbers: false,
        dateStrings: false,
        debug: true,
        trace: true,
        stringifyObjects: false,
        timezone: 'local',
        queryFormat: undefined,
        pool: undefined,
        ssl: false,
        multipleStatements: true,
        rowsAsArray: false,
        namedPlaceholders: false,
        nestTables: undefined,
        typeCast: true,
        maxPacketSize: 0,
        charsetNumber: 224,
        compress: false,
        authSwitchHandler: undefined,
        clientFlags: 259023,
        connectAttributes: undefined,
        maxPreparedStatements: 16000 },
     stream:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: null,
        _readableState: [Object],
        readable: true,
        domain: [Object],
        _events: [Object],
        _eventsCount: 5,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: false,
        destroyed: false,
        _bytesDispatched: 424,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        read: [Function],
        _consuming: true },
     _internalId: 2,
     _commands: { [String: ''] '0': undefined, _capacity: 16, _length: 0, _front: 1 },
     _command: [Circular],
     _paused: false,
     _paused_packets: { [String: ''] _capacity: 16, _length: 0, _front: 0 },
     _statements:
      LRUCache {
        max: 16000,

        'undefined/undefined/undefinedselect * from my_table where bit_prop = ?' => { value:
           PreparedStatementInfo {
             query: 'select * from my_table where bit_prop = ?',
             id: 1,
             columns: [ [Object], [Object] ],
             parameters: [ [Object] ],
             rowParser: null,
             _connection: [Circular] } }
      },
     textProtocolParsers: {},
     binaryProtocolParsers: { 'undefined/undefined/falsefalse/false/boolean/val:253:4097/bit_prop:1:4097': [Function: BinaryRow] },
     serverCapabilityFlags: 2148530175,
     authorized: true,
     sequenceId: 1,
     threadId: 4777,
     _handshakePacket:
      Handshake {
        protocolVersion: 10,
        serverVersion: '5.5.47-0ubuntu0.14.04.1',
        capabilityFlags: 2148530175,
        connectionId: 4777,
        authPluginData1: <Buffer 36 2c 7a 44 25 2a 43 50>,
        authPluginData2: <Buffer 4a 60 27 3e 44 79 56 21 42 66 5a 71>,
        characterSet: 33,
        statusFlags: 2 },
     _fatalError: null,
     _protocolError: null,
     _outOfOrderPackets: [],
     packetParser:
      PacketParser {
        buffer: [],
        bufferLength: 0,
        packetHeaderLength: 4,
        headerLen: 0,
        length: 5,
        onPacket: [Function],
        execute: [Function: executeStart] },
     serverEncoding: 'cesu8',
     connectTimeout: null,
     connectionId: 4777 },
  _commandName: 'execute',
  options:
   { isServer: undefined,
     stream: undefined,
     host: '172.17.0.2',
     port: '3306',
     localAddress: undefined,
     socketPath: undefined,
     user: 'admin',
     password: 'admin',
     passwordSha1: undefined,
     database: 'my_database',
     connectTimeout: 10000,
     insecureAuth: false,
     supportBigNumbers: false,
     bigNumberStrings: false,
     decimalNumbers: false,
     dateStrings: false,
     debug: true,
     trace: true,
     stringifyObjects: false,
     timezone: 'local',
     queryFormat: undefined,
     pool: undefined,
     ssl: false,
     multipleStatements: true,
     rowsAsArray: false,
     namedPlaceholders: false,
     nestTables: undefined,
     typeCast: true,
     maxPacketSize: 0,
     charsetNumber: 224,
     compress: false,
     authSwitchHandler: undefined,
     clientFlags: 259023,
     connectAttributes: undefined,
     maxPreparedStatements: 16000,
     sql: 'select * from my_table where bit_prop = ?',
     values: [ true ] } }
>  raw: 0233000002
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#resultsetHeader(1,,5)
        Resultset header received, expecting 2 column definition packets
 raw: 036465660762696c6c696e67086d795f7461626c65086d795f7461626c650376616c0376616c0c210060000000fd01100000003d000003
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#readField(2,,55)
 raw: 036465660762696c6c696e67086d795f7461626c65086d795f7461626c65086269745f70726f70086269745f70726f700c3f000100000001011000000005000004
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#readField(3,,65)
 raw: fe0000220009000005
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#fieldsEOF(4,EOF,9)
 raw: 0000056c6f72656d0007000006
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#row(5,maybeOK,13)
 raw: 0000037369740005000007
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#row(6,maybeOK,11)
 raw: fe00002200
Trace
    at Connection.handlePacket (/project/node_modules/mysql2/lib/connection.js:348:15)
    at PacketParser.onPacket (/project/node_modules/mysql2/lib/connection.js:87:66)
    at PacketParser.executeStart (/project/node_modules/mysql2/lib/packet_parser.js:45:12)
    at Socket.<anonymous> (/project/node_modules/mysql2/lib/connection.js:94:29)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
2 4777 ==> execute#row(7,EOF,9)
null [ BinaryRow { val: 'lorem', bit_prop: 0 },
  BinaryRow { val: 'sit', bit_prop: 0 } ] [ { catalog: 'def',
    schema: 'my_database',
    name: 'val',
    orgName: 'val',
    table: 'my_table',
    orgTable: 'my_table',
    characterSet: 33,
    columnLength: 96,
    columnType: 253,
    flags: 4097,
    decimals: 0 },
  { catalog: 'def',
    schema: 'my_database',
    name: 'bit_prop',
    orgName: 'bit_prop',
    table: 'my_table',
    orgTable: 'my_table',
    characterSet: 63,
    columnLength: 1,
    columnType: 1,
    flags: 4097,
    decimals: 0 } ]
```
</details>

**The resulting rows:**

```javascript
[ BinaryRow { val: 'lorem', bit_prop: 0 },
  BinaryRow { val: 'sit', bit_prop: 0 } ]
```

As you can see we get rows with `bit_prop = 0` when `true` is passed as parameter.

**Comment by sidorares** [comment]
`tinyint` (and boolean) are in fact 1 byte signed int type. This works as expected:

```js
conn.execute('select * from my_table where bit_prop = ?', [1], console.log)
```

what currently happens is that `true` is sent as `"true"` string, and result is the same as doing `conn.execute('select * from my_table where bit_prop = "true"')`. 

Maybe we should convert js boolean values to 0/1 when sending them as parameters, not sure.

**Comment by joakimbeng** [comment]
IMO I think you should, because that would be inline with how `query` works.

**Comment by joakimbeng** [comment]
I haven't verified but by looking at the code I think #353 solves this, because it casts booleans to the `TINY` type.

**Comment by joakimbeng** [comment]
Haha, now a couple of months later I had issues with this again and a google search brought me here! It took me a while to realize it was me who created the issue :rofl:

**Comment by sidorares** [comment]
lol :)

**Comment by Clotoo** [comment]
Running into the same issue. I am using TINYINT(1) as booleans.
I tested a bit and experienced the following behaviors :

execute("SELECT * FROM Object WHERE bool = ?", [true])
==> returns all rows

execute("UPDATE Object SET bool=1 WHERE bool = ?", [true])
==> updates all rows

execute("UPDATE Object SET bool = ? WHERE bool=1", [true])
==> crashes with Error: Incorrect integer value: 'true' for column 'bool' at row 1

execute("INSERT INTO Object SET bool = ?", [true])
==> crashes with Error: Incorrect integer value: 'true' for column 'bool' at row 1

execute("DELETE FROM Object WHERE bool = ?", [true])
==> deletes all rows

Basically when the booleans are only used in the WHERE, it can be dangerous because it seems like it works, but nope it's just like an empty condition.
When used to write values (update/create) then it fails with an error.

I am currently working around this with a small override for execute (Promise-based version) :
```
db._execute = db.execute;
db.execute = function(sql, values) {
    // Convert booleans to TINYINT(1)
    for ( var i=0; i<values.length; i++ )
        (typeof(values[i]) == 'boolean') && (values[i] = values[i] ? 1 : 0);
    return db._execute(sql, values)
}
```

**Comment by barkermn01** [comment]
how is this still open?
